File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/domain/Conversation.java
package systems.zlink.samples.supportchat.server.support.domain;

import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Conversation {
    public enum Status {
        WaitingForAgent,
        Active,
        WaitingForClose,
        Closed
    }

    public enum Role {
        Customer,
        Agent
    }

    public enum EventKind {
        ParticipantJoined,
        MessageAppended,
        TypingChanged,
        Idle,
        Closed
    }

    public record Policy(Duration idleTimeout, Duration closeGraceTimeout, int maxMessageLength) {}

    public record Participant(
            String actorId, Role role, String displayName, long joinedAtUnixMs, boolean typing) {}

    public record Message(
            String conversationId,
            long messageSeq,
            String senderActorId,
            String text,
            long sentAtUnixMs) {}

    public record Snapshot(
            String conversationId,
            String subject,
            Status status,
            String customerActorId,
            String agentActorId,
            long lastMessageSeq,
            Long lastMessageAtUnixMs,
            Long idleDeadlineUnixMs) {}

    public record Event(
            EventKind kind,
            Snapshot state,
            String actorId,
            Role role,
            Message message,
            Boolean typing) {}

    public record Change(Snapshot state, List<Event> events) {}

    private final String conversationId;
    private final String subject;
    private final String customerActorId;
    private final Policy policy;
    private final Map<String, Participant> participants = new LinkedHashMap<>();
    private Status status = Status.WaitingForAgent;
    private String agentActorId;
    private long lastMessageSeq;
    private Long lastMessageAtUnixMs;
    private Long idleDeadlineUnixMs;
    private Long closeDeadlineUnixMs;

    public Conversation(
            String conversationId,
            String subject,
            String customerActorId,
            String customerDisplayName,
            long createdAtUnixMs,
            Policy policy) {
        if (subject == null || subject.isBlank()) {
            throw new IllegalArgumentException("Conversation subject is required");
        }
        this.conversationId = conversationId;
        this.subject = subject;
        this.customerActorId = customerActorId;
        this.policy = policy;
        participants.put(
                customerActorId,
                new Participant(
                        customerActorId,
                        Role.Customer,
                        customerDisplayName,
                        createdAtUnixMs,
                        false));
    }

    public synchronized Snapshot snapshot() {
        return new Snapshot(
                conversationId,
                subject,
                status,
                customerActorId,
                agentActorId,
                lastMessageSeq,
                lastMessageAtUnixMs,
                idleDeadlineUnixMs);
    }

    public synchronized Change joinAgent(String actorId, String displayName, long now) {
        ensureOpen("join an agent");
        if (agentActorId != null && !agentActorId.equals(actorId)) {
            throw new IllegalStateException("Conversation already has an assigned agent");
        }
        agentActorId = actorId;
        status = Status.Active;
        participants.put(actorId, new Participant(actorId, Role.Agent, displayName, now, false));
        Snapshot state = snapshot();
        return change(
                new Event(EventKind.ParticipantJoined, state, actorId, Role.Agent, null, null));
    }

    public synchronized Change sendMessage(String senderActorId, String text, long now) {
        requireParticipant(senderActorId);
        if (status == Status.Closed) {
            throw new IllegalStateException("Closed conversation cannot accept messages");
        }
        if (status == Status.WaitingForAgent) {
            throw new IllegalStateException("Conversation is waiting for an agent");
        }
        if (text == null || text.isBlank()) {
            throw new IllegalArgumentException("Message text is required");
        }
        if (text.length() > policy.maxMessageLength()) {
            throw new IllegalArgumentException("Message text is too long");
        }
        status = Status.Active;
        lastMessageSeq += 1;
        lastMessageAtUnixMs = now;
        idleDeadlineUnixMs = now + policy.idleTimeout().toMillis();
        closeDeadlineUnixMs = null;
        Message message = new Message(conversationId, lastMessageSeq, senderActorId, text, now);
        Snapshot state = snapshot();
        return change(
                new Event(EventKind.MessageAppended, state, senderActorId, null, message, null));
    }

    public synchronized Change setTyping(String actorId, boolean typing) {
        Participant participant = participants.get(actorId);
        if (status == Status.Closed || participant == null) {
            return new Change(snapshot(), List.of());
        }
        participants.put(
                actorId,
                new Participant(
                        participant.actorId(),
                        participant.role(),
                        participant.displayName(),
                        participant.joinedAtUnixMs(),
                        typing));
        Snapshot state = snapshot();
        return change(
                new Event(
                        EventKind.TypingChanged, state, actorId, participant.role(), null, typing));
    }

    public synchronized Change markIdle(long now) {
        if (status == Status.Active && idleDeadlineUnixMs != null && now >= idleDeadlineUnixMs) {
            status = Status.WaitingForClose;
            closeDeadlineUnixMs = now + policy.closeGraceTimeout().toMillis();
            Snapshot state = snapshot();
            return change(new Event(EventKind.Idle, state, null, null, null, null));
        }
        if (status == Status.WaitingForClose
                && closeDeadlineUnixMs != null
                && now >= closeDeadlineUnixMs) {
            status = Status.Closed;
            Snapshot state = snapshot();
            return change(new Event(EventKind.Closed, state, null, null, null, null));
        }
        return new Change(snapshot(), List.of());
    }

    public synchronized Change close(String actorId, String reason) {
        requireParticipant(actorId);
        if (status == Status.Closed) {
            throw new IllegalStateException("Conversation is already closed");
        }
        status = Status.Closed;
        Snapshot state = snapshot();
        return change(new Event(EventKind.Closed, state, actorId, null, null, null));
    }

    private Participant requireParticipant(String actorId) {
        Participant participant = participants.get(actorId);
        if (participant == null) {
            throw new IllegalStateException("Actor is not a conversation participant");
        }
        return participant;
    }

    private void ensureOpen(String action) {
        if (status == Status.Closed) {
            throw new IllegalStateException("Closed conversation cannot " + action);
        }
    }

    private Change change(Event event) {
        return new Change(event.state(), List.of(event));
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/domain/Conversation.kt
package systems.zlink.samples.kotlin.supportchat.server.support.domain

class Conversation(
    val conversationId: String,
    val subject: String,
    val customerActorId: String,
    customerDisplayName: String,
    createdAtUnixMs: Long,
    private val policy: ConversationPolicy = ConversationPolicy.Sample,
) {
    private val messages = mutableListOf<ConversationMessage>()
    private val participants = linkedMapOf<String, ConversationParticipant>()
    private var closeDeadlineUnixMs: Long? = null
    private var idleDeadlineUnixMs: Long? = null
    private var lastMessageAtUnixMs: Long? = null
    private var lastMessageSeq: Long = 0

    var status: ConversationStatus = ConversationStatus.WaitingForAgent
        private set

    var agentActorId: String? = null
        private set

    init {
        require(subject.isNotBlank()) { "Conversation subject is required." }
        participants[customerActorId] =
            ConversationParticipant(
                actorId = customerActorId,
                role = ParticipantRole.Customer,
                displayName = customerDisplayName,
                joinedAtUnixMs = createdAtUnixMs,
                isTyping = false,
            )
    }

    fun snapshot(): ConversationSnapshot =
        ConversationSnapshot(
            conversationId = conversationId,
            subject = subject,
            status = status,
            customerActorId = customerActorId,
            agentActorId = agentActorId,
            lastMessageSeq = lastMessageSeq,
            lastMessageAtUnixMs = lastMessageAtUnixMs,
            idleDeadlineUnixMs = idleDeadlineUnixMs,
        )

    fun joinAgent(
        agentActorId: String,
        agentDisplayName: String,
        joinedAtUnixMs: Long,
    ): ConversationChange {
        ensureNotClosed("join an agent")
        val currentAgent = this.agentActorId
        check(currentAgent == null || currentAgent == agentActorId) {
            "Conversation already has an assigned agent."
        }

        this.agentActorId = agentActorId
        status = ConversationStatus.Active
        participants[agentActorId] =
            ConversationParticipant(
                actorId = agentActorId,
                role = ParticipantRole.Agent,
                displayName = agentDisplayName,
                joinedAtUnixMs = joinedAtUnixMs,
                isTyping = false,
            )

        val state = snapshot()
        return ConversationChange(
            state = state,
            events =
                listOf(
                    ConversationEvent(
                        kind = ConversationEventKind.ParticipantJoined,
                        state = state,
                        actorId = agentActorId,
                        role = ParticipantRole.Agent,
                    )
                ),
        )
    }

    fun sendMessage(senderActorId: String, text: String, sentAtUnixMs: Long): ConversationChange {
        ensureParticipant(senderActorId)
        check(status != ConversationStatus.Closed) { "Closed conversation cannot accept messages." }
        check(status != ConversationStatus.WaitingForAgent) {
            "Conversation is waiting for an agent."
        }
        require(text.isNotBlank()) { "Message text is required." }
        require(text.length <= policy.maxMessageLength) { "Message text is too long." }

        status = ConversationStatus.Active
        lastMessageSeq += 1
        lastMessageAtUnixMs = sentAtUnixMs
        idleDeadlineUnixMs = sentAtUnixMs + policy.idleTimeout.toMillis()
        closeDeadlineUnixMs = null
        val message =
            ConversationMessage(
                conversationId = conversationId,
                messageSeq = lastMessageSeq,
                senderActorId = senderActorId,
                text = text,
                sentAtUnixMs = sentAtUnixMs,
            )
        messages += message

        val state = snapshot()
        return ConversationChange(
            state = state,
            events =
                listOf(
                    ConversationEvent(
                        kind = ConversationEventKind.MessageAppended,
                        state = state,
                        actorId = senderActorId,
                        message = message,
                    )
                ),
        )
    }

    fun setTyping(actorId: String, isTyping: Boolean): ConversationChange {
        val participant = participants[actorId]
        if (status == ConversationStatus.Closed || participant == null) {
            return ConversationChange(snapshot(), emptyList())
        }

        participants[actorId] = participant.copy(isTyping = isTyping)
        val state = snapshot()
        return ConversationChange(
            state = state,
            events =
                listOf(
                    ConversationEvent(
                        kind = ConversationEventKind.TypingChanged,
                        state = state,
                        actorId = actorId,
                        role = participant.role,
                        isTyping = isTyping,
                    )
                ),
        )
    }

    fun markIdle(nowUnixMs: Long): ConversationChange {
        val idleDeadline = idleDeadlineUnixMs
        if (
            status == ConversationStatus.Active && idleDeadline != null && nowUnixMs >= idleDeadline
        ) {
            status = ConversationStatus.WaitingForClose
            closeDeadlineUnixMs = nowUnixMs + policy.closeGraceTimeout.toMillis()
            val state = snapshot()
            return ConversationChange(
                state = state,
                events = listOf(ConversationEvent(ConversationEventKind.Idle, state)),
            )
        }

        val closeDeadline = closeDeadlineUnixMs
        if (
            status == ConversationStatus.WaitingForClose &&
                closeDeadline != null &&
                nowUnixMs >= closeDeadline
        ) {
            status = ConversationStatus.Closed
            val state = snapshot()
            return ConversationChange(
                state = state,
                events = listOf(ConversationEvent(ConversationEventKind.Closed, state)),
            )
        }

        return ConversationChange(snapshot(), emptyList())
    }

    fun close(actorId: String, reason: String?): ConversationChange {
        reason?.let { _ -> }
        ensureParticipant(actorId)
        check(status != ConversationStatus.Closed) { "Conversation is already closed." }

        status = ConversationStatus.Closed
        val state = snapshot()
        return ConversationChange(
            state = state,
            events =
                listOf(
                    ConversationEvent(
                        kind = ConversationEventKind.Closed,
                        state = state,
                        actorId = actorId,
                    )
                ),
        )
    }

    private fun ensureParticipant(actorId: String): ConversationParticipant =
        participants[actorId] ?: error("Actor is not a conversation participant.")

    private fun ensureNotClosed(action: String) {
        check(status != ConversationStatus.Closed) { "Closed conversation cannot $action." }
    }
}
